cachi2/core/models/sbom.py

```diff
@@ -87,20 +86,624 @@ class Metadata(pydantic.BaseModel):
+
+        def create_document_root() -> SPDXPackage:
+            return SPDXPackage(name="", versionInfo="", SPDXID="SPDXRef-DocumentRoot-File-")
+
+        def create_root_relationship() -> SPDXRelation:
+            return SPDXRelation(
+                spdxElementId="SPDXRef-DOCUMENT",
+                comment="",
+                relatedSpdxElement="SPDXRef-DocumentRoot-File-",
+                relationshipType="DESCRIBES",
+            )
+
+        def link_to_root(packages: Union[list[SPDXPackage], list[SPDXFile]]) -> list[SPDXRelation]:
+            relationships, root_id, rtype = [], "SPDXRef-DocumentRoot-File-", "CONTAINS"
+            pRel = partial(SPDXRelation, spdxElementId=root_id, comment="", relationshipType=rtype)
+            for package in packages:
+                if package.SPDXID == "SPDXRef-DocumentRoot-File-":
+                    continue
+                relationships.append(pRel(relatedSpdxElement=package.SPDXID))
+            return relationships
+
+        def libs_to_packages(libraries: list[Component]) -> list[SPDXPackage]:
+            packages, annottr, now = [], "Tool: cachi2:jsonencoded", spdx_now()
+            args = dict(annotator=annottr, annotationDate=now, annotationType="OTHER")
+            pAnnotation = partial(SPDXPackageAnnotation, **args)
+
+            # noqa for trivial helpers.
+            mkcomm = lambda p: json.dumps(dict(name=f"{p.name}", value=f"{p.value}"))  # noqa: E731
+            hashdict = lambda c: dict(name=c.name, version=c.version, purl=c.purl)  # noqa: E731
+            erefbase = dict(referenceCategory="PACKAGE-MANAGER", referenceType="purl")
+            erefdict = lambda c: dict(referenceLocator=c.purl, **erefbase)  # noqa: E731
+
+            for component in libraries:
+                package_hash = SPDXPackage._calculate_package_hash_from_dict(hashdict(component))
+                packages.append(
+                    SPDXPackage(
+                        SPDXID=f"SPDXRef-Package-{component.name}-{component.version}-{package_hash}",
+                        name=component.name,
+                        versionInfo=component.version,
+                        externalRefs=[erefdict(component)],
+                        annotations=[pAnnotation(comment=mkcomm(p)) for p in component.properties],
+                    )
+                )
+            return packages
+
+        def files_to_packages(cydxfiles: list[Component]) -> list[SPDXFile]:
+            return [SPDXFile(SPDXID="SPDXRef-File-{f.name}", fileName=f.name) for f in cydxfiles]
+
+        # Main function body.
+        cydxfiles, libraries = partition_by(lambda c: c.type == "library", self.components)
+        # mypy is upset by patrition_by being broadly typed.
+        packages = [create_document_root()] + libs_to_packages(libraries)  # type: ignore
+        files = files_to_packages(cydxfiles)  # type: ignore
+        relationships = [create_root_relationship()] + link_to_root(packages) + link_to_root(files)
```

Comment:
It seems to me that the spec does not require these relationships to be informed (I also checked the official [examples](https://github.com/spdx/spdx-examples/tree/master) to try to figure this out). 

Now, IIUC, we're saying that every dependency is contained by this "virtual" root package `SPDXRef-DocumentRoot-File-`. I'm wondering if this is adding any value in the end, as it does add a lot of extra lines to the SBOM.

Reply:
I am following a precedent set by Syft here. Since these relationships seem to be added by default I think it is best to keep them (there will be some interaction with Syft, at least for some time).

Reply:
Apparently, Syft does at least two types of relationships:

```
{
  "spdxElementId": "SPDXRef-DocumentRoot-Directory--home-bpimente-Projects-test-repos-porta",
  "relatedSpdxElement": "SPDXRef-Package-npm-babel-plugin-polyfill-corejs3-a77653646f05b06b",
  "relationshipType": "CONTAINS"
}
```

And one for the file that described the dependency (e.g. `package.json`):

```
{
  "spdxElementId": "SPDXRef-Package-npm-string-decoder-14144cdbecdae249",
  "relatedSpdxElement": "SPDXRef-File-yarn.lock-dcee26e98fae7b36",
  "relationshipType": "OTHER",
  "comment": "evident-by: indicates the package's existence is evident by the given file"
}
```

I'm not sure if we'd also want to add the latter at some point, but it does feel more useful than the first one.

Reply:
In any case, feel free to close this thread, it's a non-blocker.